src/main/java/edu/umich/tl/analytics/CanvasCourseReportServlet.java

```diff
@@ -74,32 +84,121 @@ protected void doGet(HttpServletRequest request,HttpServletResponse response){
 		}
 		
 	}
+
+	private void checkIfACourseReportThreadIsDone(HttpServletRequest request,HttpServletResponse response) {
+		M_log.debug("Starting polling... Thread size: "+courseReportReqThreads.size());
+		try {
+			String thrdId = request.getParameter(THREAD_ID);
+			long threadId = Long.parseLong(thrdId);
+			KeySetView<Thread, CourseReportTask> keySet = courseReportReqThreads.keySet();
+			for (Thread aCourseReportThread : keySet) {
+				if(aCourseReportThread.getId()==threadId) {
+					checkIfThreadAlive(response, aCourseReportThread);
+					break;
+				}
+
+			}
+		} catch (IOException e) {
+			M_log.debug("IOException occoured in checkIfACourseReportThreadIsDone()"+e);
+			response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+			try {
+				response.getWriter().print(props.getString("report.generation.err.msg"));
+			} catch (IOException e1) {
+				M_log.error("IOException occurred while writing the response out",e1);
+			}
+		}
+	}
+
+
+	private void checkIfThreadAlive(HttpServletResponse response, Thread aCourseReportThread)
+			throws IOException {
+		OutputStream outputStream = response.getOutputStream();
+		if(aCourseReportThread.isAlive()) {
+			outputStream.write(WORKING.getBytes());
+			outputStream.flush();
+			outputStream.close();
+		}else {
+			afterActionThreadIsDone(response, aCourseReportThread, outputStream);
+		}
+	}
+
+
+	private void afterActionThreadIsDone(HttpServletResponse response, Thread aCourseReportThread,
+			OutputStream outputStream)
```

Comment:
It's not necessary to add `outputStream` to the arguments of this method, when that stream can easily be gotten from `response`.  Too many method arguments makes the code more difficult to read.  Alternatively, if you really don't want to get the stream from the response object, you could store it as an attribute of the class and have `setOutputStream()` and `getOutputStream()` methods.